File: examples/sorting-algorithms/sut-insertion-sort.ts
/**
 * Insertion Sort SUT
 *
 * Builds a sorted prefix by shifting elements rightward to insert
 * each new element in its correct position. Each shift counts as a swap.
 */

interface SortInput {
	data: number[];
}

interface SortResult {
	sorted: number[];
	comparisons: number;
	swaps: number;
	executionTimeMs: number;
}

export function createSut(): {
	id: string;
	config: Record<string, unknown>;
	run: (input: SortInput) => Promise<SortResult>;
} {
	return {
		id: "insertion-sort",
		config: {},
		run: async (input: SortInput): Promise<SortResult> => {
			const arr = [...input.data];
			let comparisons = 0;
			let swaps = 0;

			const start = performance.now();

			for (let i = 1; i < arr.length; i++) {
				const key = arr[i];
				let j = i - 1;

				while (j >= 0) {
					comparisons++;
					if (arr[j] > key) {
						arr[j + 1] = arr[j];
						swaps++;
						j--;
					} else {
						break;
					}
				}
				arr[j + 1] = key;
			}

			const executionTimeMs = performance.now() - start;

			return { sorted: arr, comparisons, swaps, executionTimeMs };
		},
	};
}

File: examples/sorting-algorithms/sut_insertion_sort.py
"""Insertion Sort SUT — shift-insert pattern."""

import time


def create_sut():
    def run(input):
        arr = list(input["data"])
        comparisons = 0
        swaps = 0

        start = time.perf_counter()

        for i in range(1, len(arr)):
            key = arr[i]
            j = i - 1

            while j >= 0:
                comparisons += 1
                if arr[j] > key:
                    arr[j + 1] = arr[j]
                    swaps += 1
                    j -= 1
                else:
                    break
            arr[j + 1] = key

        execution_time_ms = (time.perf_counter() - start) * 1000

        return {
            "sorted": arr,
            "comparisons": comparisons,
            "swaps": swaps,
            "executionTimeMs": execution_time_ms,
        }

    return {"id": "insertion-sort", "config": {}, "run": run}
